src/pubsub/mod_pubsub.erl

```diff
@@ -1268,40 +1270,86 @@ iq_pubsub(Host, ServerHost, From, IQType, QueryEl, Lang) ->
 iq_pubsub(Host, ServerHost, From, IQType, #xmlel{children = SubEls} = QueryEl,
           Lang, Access, Plugins) ->
     case xml:remove_cdata(SubEls) of
-        [#xmlel{name = Name, attrs = ActionAttrs, children = ActionSubEls} = ActionEl | _] ->
-            Node = xml:get_attr_s(<<"node">>, ActionAttrs),
-            case {IQType, Name} of
-                {set, <<"create">>} ->
-                    iq_pubsub_set_create(Host, ServerHost, Node, From, Access,
-                                         Plugins, ActionEl, QueryEl);
-                {set, <<"publish">>} ->
-                    iq_pubsub_set_publish(Host, ServerHost, Node, From, Access,
-                                         ActionSubEls, QueryEl);
-                {set, <<"retract">>} ->
-                    iq_pubsub_set_retract(Host, Node, From, ActionAttrs, ActionSubEls);
-                {set, <<"subscribe">>} ->
-                    iq_pubsub_set_subscribe(Host, Node, From, QueryEl, ActionAttrs);
-                {set, <<"unsubscribe">>} ->
-                    iq_pubsub_set_unsubscribe(Host, Node, From, ActionAttrs);
-                {get, <<"items">>} ->
-                    iq_pubsub_get_items(Host, Node, From, QueryEl, ActionAttrs, ActionSubEls);
-                {get, <<"subscriptions">>} ->
-                    get_subscriptions(Host, Node, From, Plugins);
-                {get, <<"affiliations">>} ->
-                    get_affiliations(Host, Node, From, Plugins);
-                {get, <<"options">>} ->
-                    iq_pubsub_get_options(Host, Node, Lang, ActionAttrs);
-                {set, <<"options">>} ->
-                    XEl = exml_query:subelement_with_name_and_ns(ActionEl, <<"x">>, ?NS_XDATA),
-                    iq_pubsub_set_options(Host, Node, ActionAttrs, XEl);
-                _ ->
-                    {error, mongoose_xmpp_errors:feature_not_implemented()}
-            end;
+        [#xmlel{name = Name} = ActionEl | _] ->
+            report_iq_action_metrics_before_result(ServerHost, IQType, Name),
+            Node = exml_query:attr(ActionEl, <<"node">>, <<>>),
+            {Time, Result} = timer:tc(fun iq_pubsub_action/11,
+                                      [IQType, Name, Host, ServerHost,
+                                       Plugins, Node, From, Access, ActionEl,
```

Comment:
Maybe it makes sense to create a map here with all params and pass it to all subsequent functions without extracting them before they are needed? I think it won't be that invasive piece of refactoring.

Reply:
Sure, I'll try to do that.